### PR TITLE
istiod mwc enableNamespacesByDefault should support custom namespaces

### PR DESCRIPTION
**Pull request by tariq1890**
We would like to use this setting `enableNamespacesByDefault`, but this turns off istio-injection for the Namespace the chart belongs to. Since we're using a custom namespace to house our Istio deployments and also injecting sidecars into the same, we need the helm charts to support this use case

[ ] Configuration Infrastructure
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by costinm — COMMENTED** [review]
I don't get it - hardcoding istio-system doesn't seem right. What if you install the injector and istiod in another namespace ? 

Don't we already configure the webhook to ignore pods based on annotation ? Istiod should have the annotation ( and gateway should not since we're planning to start injecting it )

**Comment by tariq1890** [comment]
>I don't get it - hardcoding istio-system doesn't seem right. What if you install the injector and istiod in another namespace ?

This is exactly why I have this PR. In our case, we install istio in a custom namespace, but we also need injection to be enabled in that same custom namespace. The current helm charts will not allow this

**Comment by tariq1890** [comment]
@costinm I have removed the hardcoded `istio-system` as per your suggestion.

File: manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml [hunk 1]
```diff
@@ -37,10 +37,12 @@ webhooks:
     namespaceSelector:
 {{- if .Values.sidecarInjectorWebhook.enableNamespacesByDefault }}
       matchExpressions:
+      {{- if .Values.sidecarInjectorWebhook.excludeNamespaces }}
       - key: name
         operator: NotIn
         values:
-        - {{ .Release.Namespace }}
+        {{ toYaml .Values.sidecarInjectorWebhook.excludeNamespaces | indent 8 | trim }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:44]: isn't this a breaking change? We no longer will exclude the release namespace which may cause users istio control plane pods to be injected on upgrade and break?

Comment (tariq1890) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:44]: Looks like some of the components have the annotation, but not the labels. I can add the labels and annotation of no-inject where necessary (like cni, gateways, istiod etc)

**Comment by tariq1890** [comment]
@howardjohn @costinm I have addressed your review comments. Ready for another round of review

**Comment by tariq1890** [comment]
@costinm @howardjohn requesting your review on this

File: manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml [hunk 2]
```diff
@@ -37,10 +37,17 @@ webhooks:
     namespaceSelector:
 {{- if .Values.sidecarInjectorWebhook.enableNamespacesByDefault }}
       matchExpressions:
+      {{- if .Values.sidecarInjectorWebhook.excludeNamespaces }}
+      - key: name
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:41]: Just to be sure this is known, `name` is NOT the name of the namespace. It is a `name` label, which is not automatically added (and doesn't follow k8s naming practices!). There is some KEP to automatically add a similar label, but its not ready yet

**Review by howardjohn — COMMENTED** [review]
Looks good to me

File: manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml (hunk 2 above)
Comment (tariq1890) [reply manifests/charts/istio-control/istio-discovery/templates/mutatingwebhook.yaml:41]: Thanks for the info @howardjohn. This is good to know. Where is the KEP for this?

**Comment by howardjohn** [comment]
https://github.com/kubernetes/enhancements/pull/2162

**Pull request closed without merging**